lnwire/onion_error.go

```diff
@@ -294,49 +294,73 @@ func (f FailUnknownNextPeer) Error() string {
 	return f.Code().String()
 }
 
-// FailUnknownPaymentHash is returned If the payment hash has already been
-// paid, the final node MAY treat the payment hash as unknown, or may succeed
-// in accepting the HTLC. If the payment hash is unknown, the final node MUST
-// fail the HTLC.
+// FailIncorrectPaymentAmount is returned if the amount paid is less than the
+// amount expected, the final node MUST fail the HTLC. If the amount paid is
+// more than twice the amount expected, the final node SHOULD fail the HTLC.
+// This allows the sender to reduce information leakage by altering the amount,
+// without allowing accidental gross overpayment.
 //
 // NOTE: May only be returned by the final node in the path.
-type FailUnknownPaymentHash struct{}
+type FailIncorrectPaymentAmount struct{}
 
 // Code returns the failure unique code.
 //
 // NOTE: Part of the FailureMessage interface.
-func (f FailUnknownPaymentHash) Code() FailCode {
-	return CodeUnknownPaymentHash
+func (f FailIncorrectPaymentAmount) Code() FailCode {
+	return CodeIncorrectPaymentAmount
 }
 
 // Returns a human readable string describing the target FailureMessage.
 //
 // NOTE: Implements the error interface.
-func (f FailUnknownPaymentHash) Error() string {
+func (f FailIncorrectPaymentAmount) Error() string {
 	return f.Code().String()
 }
 
-// FailIncorrectPaymentAmount is returned if the amount paid is less than the
-// amount expected, the final node MUST fail the HTLC. If the amount paid is
-// more than twice the amount expected, the final node SHOULD fail the HTLC.
-// This allows the sender to reduce information leakage by altering the amount,
-// without allowing accidental gross overpayment.
+// FailUnknownPaymentHash is returned for two reasons:
+//
+// 1) if the payment hash has already been paid, the final node MAY treat the
+// payment hash as unknown, or may succeed in accepting the HTLC. If the
+// payment hash is unknown, the final node MUST fail the HTLC.
+//
+// 2) if the amount paid is less than the amount expected, the final node MUST
+// fail the HTLC. If the amount paid is more than twice the amount expected,
+// the final node SHOULD fail the HTLC. This allows the sender to reduce
+// information leakage by altering the amount, without allowing accidental
+// gross overpayment.
 //
 // NOTE: May only be returned by the final node in the path.
-type FailIncorrectPaymentAmount struct{}
+type FailUnknownPaymentHash struct {
```

Comment:
we should consider adding a constructor for this error, and migrate all uses of this error in the code base to use it. this should help mitigate the chance of errors introduced by not setting `Amount`. In addition, I think it'd be worth making the amount private, so that it can only be accessed via `Amount()`

Reply:
Fixed in latest diff!